hlsdl.go

```diff
@@ -179,67 +157,121 @@ func (hlsDl *HlsDl) downloadSegments(segments []*Segment) error {
 				close(quitChan)
 				return result.Err
 			}
-
 			if hlsDl.enableBar {
 				hlsDl.bar.Increment()
+			} else {
+				atomic.AddInt64(&hlsDl.segCurrent, 1)
 			}
 		}
 	}
 
 }
 
-func (hlsDl *HlsDl) join(dir string, segments []*Segment) (string, error) {
+func (hlsDl *HlsDl) join(segmentsDir string, segments []*Segment) (string, error) {
 	log.Println("Joining segments")
 
-	filepath := filepath.Join(dir, hlsDl.filename)
+	outFile := filepath.Join(hlsDl.dir, hlsDl.filename)
 
-	file, err := os.Create(filepath)
+	f, err := os.Create(outFile)
 	if err != nil {
 		return "", err
 	}
-	defer file.Close()
+	defer f.Close()
 
 	sort.Slice(segments, func(i, j int) bool {
 		return segments[i].SeqId < segments[j].SeqId
 	})
 
+	defer os.RemoveAll(segmentsDir)
 	for _, segment := range segments {
-
 		d, err := hlsDl.decrypt(segment)
 		if err != nil {
 			return "", err
 		}
-
-		if _, err := file.Write(d); err != nil {
+		if _, err := f.Write(d); err != nil {
 			return "", err
 		}
-
 		if err := os.RemoveAll(segment.Path); err != nil {
 			return "", err
 		}
 	}
 
-	return filepath, nil
+	return outFile, nil
 }
 
 func (hlsDl *HlsDl) Download() (string, error) {
 	segs, err := parseHlsSegments(hlsDl.hlsURL, hlsDl.headers)
 	if err != nil {
 		return "", err
 	}
-
-	if err := os.MkdirAll(hlsDl.dir, os.ModePerm); err != nil {
+	segmentsDir := filepath.Join(hlsDl.dir, fmt.Sprintf("%d", hlsDl.startTime))
+	if err := os.MkdirAll(segmentsDir, os.ModePerm); err != nil {
 		return "", err
 	}
-
-	if err := hlsDl.downloadSegments(segs); err != nil {
+	if err := hlsDl.downloadSegments(segmentsDir, segs); err != nil {
 		return "", err
 	}
-
-	filepath, err := hlsDl.join(hlsDl.dir, segs)
+	fp, err := hlsDl.join(segmentsDir, segs)
 	if err != nil {
 		return "", err
 	}
 
-	return filepath, nil
+	return fp, nil
+}
+
+// Decrypt descryps a segment
+func (hlsDl *HlsDl) decrypt(segment *Segment) ([]byte, error) {
+	file, err := os.Open(segment.Path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+	data, err := io.ReadAll(file)
+	if err != nil {
+		return nil, err
+	}
+	if segment.Key != nil {
+		key, iv, err := hlsDl.getKey(segment)
+		if err != nil {
+			return nil, err
+		}
+		data, err = decryptAES128(data, key, iv)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for j := 0; j < len(data); j++ {
+		if data[j] == syncByte {
+			data = data[j:]
+			break
+		}
+	}
+
+	return data, nil
+}
+
+func (hlsDl *HlsDl) getKey(segment *Segment) (key []byte, iv []byte, err error) {
+	res, err := hlsDl.client.SetHeaders(hlsDl.headers).R().Get(segment.Key.URI)
+	if err != nil {
+		return nil, nil, err
+	}
+	if res.StatusCode() != 200 {
+		return nil, nil, errors.New("Failed to get descryption key")
+	}
+	key = res.Body()
+	iv = []byte(segment.Key.IV)
+	if len(iv) == 0 {
+		iv = defaultIV(segment.SeqId)
+	}
+	return
+}
+func (hlsDl *HlsDl) GetProgress() float64 {
```

Comment:
@yukunwu 

Showing progress is usually a real-time task. I think this function seems not to be fashionable. I suggest opening a channel that allows others to subscribe to listen to the progress's changes. What do you think?

Reply:
I agree with you.I will remove this function.We can add this in next version.